File: ruby/5.rb
#!/usr/bin/env ruby

require 'timing_method'

# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
#
# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

def gcd(a, b)
  while b != 0
    t = b
    b = (a%b)
    a = t
  end
  a
end

# lcm = product of x*y divided by the gcd of x & y
def lcm(x, y)
  (x*y) / gcd(x, y)
end

# My personal method
timing_method do
  result = 1
  a = 1

  max = 20

  # (1..max).each would also work.
  (1..max).each{|i|
    result = lcm(i, result)
  }

  puts "The least common multiple of all numbers up to #{max} is #{result}"
end

File: c/5.c
/*
 * =====================================================================================
 *
 *       Filename:  5.c
 *
 *    Description:  Problem 5
 *
 *        Version:  1.0
 *        Created:  08/15/2011 16:49:32
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Matt Simpson (Maui Threv), mr.threv@gmail.com
 *        Company:
 *
 * =====================================================================================
 */

#include <stdlib.h>
#include <stdio.h>

/*
 * ===  FUNCTION  ======================================================================
 *         Name:  gcd
 *  Description:  Greatest Common Divisor
 * =====================================================================================
 */
int gcd (int x, int y )
{
  int t = 0;
  while ( y != 0 ) {
    t = y;
    y = (x%y);
    x = t;
  }
  return x;
}

/*
 * ===  FUNCTION  ======================================================================
 *         Name:  lcm
 *  Description:  Least Common Multiple
 * =====================================================================================
 */
double lcm ( int x, int y )
{
  return (x * y) / gcd(x, y);
}		/* -----  end of function lcm  ----- */

/*
 * ===  FUNCTION  ======================================================================
 *         Name:  main
 *  Description:
 * =====================================================================================
 */
int main ( int argc, char *argv[] )
{
  int i = 1;
  double result = 1;
  int max = 20;

  for (i = 1; i < max; i++) {
    result = lcm(i, result);
    printf ("%d, %.0f\n", i, result);
  }

  printf("The smallest positive number that is evenly divisible by all integers from 1 to %d is %.0f.\n", max, result);
  return EXIT_SUCCESS;
}
